packages/langium/src/workspace/configuration.ts

```diff
@@ -0,0 +1,94 @@
+    */
+    getConfiguration(language: string, configuration: string): Promise<any>;
+
+    /**
+     *  Updates the cached configurations using the `change` notification parameters.
+     *
+     * @param change The parameters of a change configuration notification.
+     * `settings` property of the change object could be expressed as `Record<string, Record<string, any>>`
+     */
+    updateConfiguration(change: DidChangeConfigurationParams): void;
+}
+
+export class DefaultConfigurationProvider implements ConfigurationProvider {
+
+    protected settings: Record<string, Record<string, any>> = {};
+    protected workspaceConfig = false;
+    protected initialized = false;
+    protected readonly sharedServices: LangiumSharedServices;
+
+    constructor(services: LangiumSharedServices) {
+        this.sharedServices = services;
+        services.lsp.LanguageServer.onInitialize(params => {
+            this.workspaceConfig = params.capabilities.workspace?.configuration ?? false;
+        });
+        services.lsp.LanguageServer.onInitialized(_params => {
+            const languages = this.sharedServices.ServiceRegistry.all;
+            services.lsp.Connection?.client.register(DidChangeConfigurationNotification.type, {
+                // Listen to configuration changes for all languages
+                section: languages.map(lang => this.toSectionName(lang.LanguageMetaData.languageId))
+            });
+        });
+    }
+
+    protected async initialize(): Promise<void> {
+        const connection = this.sharedServices.lsp.Connection;
+        if (this.workspaceConfig && connection) {
+            const languages = this.sharedServices.ServiceRegistry.all;
+            const configToUpdate: ConfigurationItem[] = languages.map(lang => { return { section: this.toSectionName(lang.LanguageMetaData.languageId) }; });
+            // get workspace configurations (default scope URI)
+            const configs = await connection.workspace.getConfiguration(configToUpdate);
+            configToUpdate.forEach((conf, idx) => {
+                this.updateSectionConfiguration(conf.section!, configs[idx]);
+            });
+        }
+        this.initialized = true;
+    }
+
+    updateConfiguration(change: DidChangeConfigurationParams): void {
+        if(!change.settings) {
+            return;
+        }
+        Object.keys(change.settings).forEach(section => {
+            this.updateSectionConfiguration(section, change.settings[section]);
+        });
+    }
+
+    protected updateSectionConfiguration(section: string, configuration: any): void {
+        this.settings[section] = configuration;
+    }
+
+    async getConfiguration(language: string, configuration: string): Promise<any> {
+        if (!this.initialized) {
+            await this.initialize();
+        }
+        const sectionName = this.toSectionName(language);
+        if (!this.settings[sectionName]) {
+            this.settings[sectionName] = {};
```

Comment:
Instead of initializing all languages, could we initialize only the requested language here?

Reply:
Changed. Thanks!